### PR TITLE
Change `verifiable_credential` to type `Vec<CRED>` in `Presentation`

### PR DESCRIPTION
**Pull request by nanderstabel**
# Description of change
This change ensures that `verifiable_credential` in `Presentation` will serialize into a `[T]` instead of a `T`, which was previously the case when `verifiable_credential` would only hold one single credential.

This previous serialization is in line with how `credentialSubject` can be serialized: https://www.w3.org/TR/vc-data-model/#credential-subject, however this is not explicitly stated for `verifiableCredential`: https://www.w3.org/TR/vc-data-model/#presentations-0.
The following examples that are used in the specification seem to imply that 'single credentials' should be serialized as `[<credential>]`:
* https://www.w3.org/TR/vc-data-model/#example-a-simple-example-of-a-verifiable-presentation
* https://www.w3.org/TR/vc-data-model/#example-usage-of-the-termsofuse-property-by-a-holder
* https://www.w3.org/TR/vc-data-model/#example-a-verifiable-presentation-that-supports-cl-signatures

The W3C VC Data Model 2.0 specification fortunately is more specific in this regard: https://www.w3.org/TR/vc-data-model-2.0/#presentations-0:
> **verifiableCredential**
    The verifiableCredential [property](https://www.w3.org/TR/vc-data-model-2.0/#dfn-property) MAY be present. The value MUST be an array of one or more [verifiable credentials](https://www.w3.org/TR/vc-data-model-2.0/#dfn-verifiable-credential), or of data derived from [verifiable credentials](https://www.w3.org/TR/vc-data-model-2.0/#dfn-verifiable-credential) in a cryptographically [verifiable](https://www.w3.org/TR/vc-data-model-2.0/#dfn-verify) format.

Therefore the introduction of the change in this PR.

Important detail: 
Both W3C VC Data Model versions state that: `MAY be present` and `**If** present...`, which imply that `verifiableCredential` is an optional property. I fail to think of any use case where one would have a presentation without any credentials included but this optional status of `verifiableCredential` is included in this PR as well.

## Links to any relevant issues
n/a

## Type of change
Add an `x` to the boxes that are relevant to your changes.

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [ ] Enhancement (a non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Fix

## How the change has been tested
* Added unit test: `test_presentation_deserialization` which tests deserializing from an example verifiable presentation (with some minor adjustments): https://www.w3.org/TR/vc-data-model/#example-a-simple-example-of-a-verifiable-presentation
* Added unit test: `test_presentation_deserialization_without_credentials` which tests deserializing a verifiable presentation as well, but one without any `verifiableCredential`.
* Added unit test: `test_presentation_builder_valid_without_credentials` which tests building a Presentation without any credentials set.

## Change checklist
Add an `x` to the boxes that are relevant to your changes.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: identity_credential/src/presentation/presentation.rs
Comment (PhilippGackstatter) [inline identity_credential/src/presentation/presentation.rs:43]: Thanks for opening the PR!

With this change we would have at least these differences to before:
1. When deserializing, an empty array in JSON would be allowed.
2. When constructing a presentation, an empty vec would be allowed to be set, which it was previously not.

Case 2 is fine, because that's our "new way" to serialize a presentation without the `verifiableCredential` field, which we probably have no use case for, but can allow without harm.

Case 1 I think we should disallow with a `deserialize_with = "path"`. We implement the VC Data Model 1.1 and it says:
> If present, the value of the verifiableCredential [property](https://www.w3.org/TR/vc-data-model/#dfn-property) MUST be constructed from one or more [verifiable credentials](https://www.w3.org/TR/vc-data-model/#dfn-verifiable-credentials)

I interpret that as: If the field is present, then it must be a non-empty array.

@nanderstabel Do you have any thoughts on that?

Comment (nanderstabel) [reply identity_credential/src/presentation/presentation.rs:43]: Yes I agree with you! Disallowing an empty array value is more in line with the spec so I took you suggestion and added
`deserialize_verifiable_credential`. 

I also added a new corresponding `Error` variant: `EmptyVerifiableCredentialsArray`. Not sure whether it adheres to any
naming convention you may have for the different Errors so please let me know if you would like me to change it!

File: identity_credential/src/error.rs
Comment (PhilippGackstatter) [inline identity_credential/src/error.rs:51]:
```suggestion
  #[error("empty verifiableCredential array in presentation")]
  EmptyVerifiableCredentialArray,
```

Nit: Would clarify where the error occured (in the presentation) and loose the `s` since the field doesn't have it :)

File: identity_credential/src/presentation/presentation.rs
Comment (PhilippGackstatter) [inline identity_credential/src/presentation/presentation.rs:231]: Since we have a dedicated error type for this case, it would be nice to check for it instead of the less precise `is_err`. If the deserialization fails for any other reason in the future than the empty array, it would still pass.

**Review by PhilippGackstatter — APPROVED** [review]
Looks good to me. Thanks again :pray:.

I'll approve this with additional comments to mean that I leave it to your judgement whether to implement them or not.

**Review by abdulmth — APPROVED** [review]
Looks good to me, thanks for putting  in the effort.

File: identity_credential/src/error.rs
Comment (nanderstabel) [reply identity_credential/src/error.rs]: I've adjusted the test to test for the specific error. I feel there must be a prettier way to do it but doing a string compare does the job as well.

File: identity_credential/src/presentation/presentation.rs
Comment (PhilippGackstatter) [inline identity_credential/src/presentation/presentation.rs:248]:
```suggestion
    assert_eq!(
      Presentation::<()>::from_json_value(json!({
        "@context": [
          "https://www.w3.org/2018/credentials/v1",
          "https://www.w3.org/2018/credentials/examples/v1"
        ],
        "holder": "did:test:abc1",
        "type": "VerifiablePresentation",
        "verifiableCredential": []
      }))
      .unwrap_err()
      .source()
      .unwrap()
      .to_string(),
      crate::Error::EmptyVerifiableCredentialArray.to_string()
    );
```

I think the serde error erases the type, right? So there's probably no better way to test.

Except you could compare against the string repr of the error, that would be more resilient.

**Review by PhilippGackstatter — COMMENTED** [review]
The Wasm examples CI workflow ran for 2:50h but didn't finish until then so I cancelled it. Had this before at some point but a rerun did the trick back then. I haven't restarted in case you want to fix the comment, otherwise just rerun the workflow.

File: identity_credential/src/presentation/presentation.rs
Comment (nanderstabel) [reply identity_credential/src/presentation/presentation.rs]: ahh yeah indeed, thanks for the suggestion :)

**Comment by nanderstabel** [comment]
@PhilippGackstatter @abdulmth thank you both for reviewing! I made the last suggested change to the unit test. Hopefully the CI workflow won't have any issues :)

**Comment by PhilippGackstatter** [comment]
@nanderstabel Our convention is to let the PR creator hit the merge button, so feel free to go ahead whenever you want. Thanks again :rocket:!

**Comment by nanderstabel** [comment]
Great, thanks again!

**Pull request merged**
